src/integrations.py

```diff
@@ -0,0 +1,233 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import json
+import logging
+from dataclasses import dataclass
+from typing import Any, Mapping
+from urllib.parse import urlparse
+
+from charms.hydra.v0.hydra_endpoints import HydraEndpointsRequirer
+from charms.kratos.v0.kratos_info import KratosInfoRequirer
+from charms.oathkeeper.v0.oathkeeper_info import OathkeeperInfoRequirer
+from charms.openfga_k8s.v1.openfga import OpenFGARequires
+from charms.tempo_k8s.v0.tracing import TracingEndpointRequirer
+from ops.model import Model
+
+from constants import PEER_INTEGRATION_NAME
+from env_vars import EnvVars
+
+logger = logging.getLogger(__name__)
+
+
+class PeerData:
+    def __init__(self, model: Model) -> None:
+        self._model = model
+        self._app = model.app
+
+    def __getitem__(self, key: str) -> dict:
+        if not (peers := self._model.get_relation(PEER_INTEGRATION_NAME)):
+            return {}
+
+        value = peers.data[self._app].get(key)
+        return json.loads(value) if value else {}
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        if not (peers := self._model.get_relation(PEER_INTEGRATION_NAME)):
+            return
+
+        peers.data[self._app][key] = json.dumps(value)
+
+    def pop(self, key: str) -> dict:
+        if not (peers := self._model.get_relation(PEER_INTEGRATION_NAME)):
+            return {}
+
+        data = peers.data[self._app].pop(key, None)
+        return json.loads(data) if data else {}
+
+
+@dataclass(frozen=True)
+class OpenFGAModelData:
+    model_id: str = ""
+
+    def to_env_vars(self) -> EnvVars:
+        return {
+            "OPENFGA_AUTHORIZATION_MODEL_ID": self.model_id,
+        }
+
+    @classmethod
+    def load(cls, source: Mapping[str, Any]) -> "OpenFGAModelData":
```

Comment:
do you mean this as an annotation `"OpenFGAModelData"` or a type hint `OpenFGAModelData`?

```suggestion
    def load(cls, source: Mapping[str, Any]) -> OpenFGAModelData:
```

Reply:
I don't quite understand the question. Before 3.10, double quotes are needed I think. I guess we need to support 3.8+? (correct me if I'm wrong).

Reply:
Sorry for the huge delay, I missed your reply.
From Python 3.8 type hints are supported so we should be good to go with real type hints ([link](https://docs.python.org/3.8/library/typing.html)).
Also, the "double quote" version is the first time I see used in our charms (but I'm no expert charmer 😄 )